### DISCUSSION TITLE
Offline Messaging System by ChatGPT

### DISCUSSION THREAD

**Discussion by Antrare** [evennia/evennia#3054] [Community Contribs & Snippets]
Here's the code for the messaging system that ChatGPT built (with a bit of bug fixing work by me) [message.py](https://github.com/evennia/evennia/files/10317399/message.py.txt)

It also suggested adding this into at_post_puppet (on character)
```python
        # Check for new messages in the player's mailbox
        mailbox = self.db.mailbox
        if mailbox:
            self.msg("You have new messages in your mailbox, use MESSAGES to view.")
```
Add the commandset to your players and it (appears) to be good to go.

**Comment by InspectorCaracal** [comment]
This is fun! I can definitely see that it was drawing from the documentation and existing code for how to structure this and it's impressively close.

That said, a warning to anyone who grabs the code: it's not a very *good* implementation. It's a reasonable starting point for anyone who wants to pre-populate the "boring" parts, but there are a lot of technical flaws.

**Reply by Antrare** [reply]
What are the technical flaws you see just running your eye over it?  I'm not pro (obviously) and keen to learn.

**Reply by InspectorCaracal** [reply]
I haven't done a detailed review yet but the things that jumped out at me:

```py
        if player.is_connected:
            # Send the message to the target player
            player.msg("Message from {}: {}".format(self.caller.name, message))
            self.caller.msg("Message sent to {}.".format(player.name))
```
This means that mail/messages won't be saved to your mailbox at all unless you're offline.

```py
            mailbox = player.db.mailbox
            if not mailbox:
                player.db.mailbox = []
            # Store the message in the target player's mailbox
            player.db.mailbox.append((self.caller, message))
```
This is not going to cause any bugs but it's using extra lines without actually saving you any db calls

The `ViewMessagesCommand` prints all of your messages all at once in one whole list with no paging. If you have a lot of messages, this is a lot of text!

There's no confirmation before deleting any of your messages. This is pretty risky! What if you put in the wrong ID?

On rereading the code to copy/paste these, I found a couple other things that made me frown:
```py
                if message_number < 1 or message_number > len(mailbox):
                    raise ValueError
```
There are much better ways of cancelling it there besides raising an error.

```py
        players = self.caller.search(player, global_search=True, quiet=True)
```
There's not really any reason to be passing quiet=True and coding your own multimatch handling here instead of using the gamewide multimatch settings

```py
        # Split the arguments into the target player and the message
        args = self.args.strip().split('=', 1)
        if len(args) != 2:
```
Evennia already has `MuxCommand` with all the nice parsing and splitting features, there's no reason to be implementing it directly here.

**Reply by InspectorCaracal** [reply]
ALL OF THAT SAID:

I'm impressed by a number of things in this output!
- It has nice comments, which I credit the source material for.
- It also correctly increments the list index by 1 for display, and decrements the input number by 1 to get the list index.
- It correctly tracks that the contents of `player.db.messages` is a list of two-item tuples.
- It added all of the command classes to the cmdset
- The `at_post_puppet` suggestion is very nice.

**Comment by InspectorCaracal** [comment]
I also want to note that while I haven't reviewed it myself, I would be very tempted to bet real money that it's cribbing from https://github.com/evennia/evennia/tree/main/evennia/contrib/game_systems/mail

***edit: I have actually looked over the contrib now and I take it back - this is a very different system! Good job, GPT, you wrote your own thing. xD

**Reply by Antrare** [reply]
It doesn't seem to do as well with scripts as it did with this, I had a lot more stuff to fix up.  But as a base/getting started, it's pretty impressive.

**Comment by Griatch** [comment]
This is very cool indeed :)
